ipa-core/src/protocol/hybrid/shuffle.rs

```diff
@@ -0,0 +1,248 @@
+        test_executor::run,
+        test_fixture::{hybrid::TestIndistinguishableHybridReport, TestWorld},
+    };
+
+    impl<BK, V> IndistinguishableHybridShare<BK, V>
+    where
+        BK: SharedValue,
+        V: SharedValue,
+    {
+        pub const ZERO: Self = Self {
+            match_key: BA64::ZERO,
+            value: V::ZERO,
+            breakdown_key: BK::ZERO,
+        };
+    }
+
+    #[tokio::test]
+    async fn hybrid_serialize_deserialize() {
+        let world = TestWorld::default();
+        let mut rng = world.rng();
+        let report = IndistinguishableHybridReport::<BA8, BA3> {
+            match_key: Replicated::<BA64>::new(rng.gen(), rng.gen()),
+            breakdown_key: Replicated::<BA8>::new(rng.gen(), rng.gen()),
+            value: Replicated::<BA3>::new(rng.gen(), rng.gen()),
+        };
+
+        let left = report.left();
+
+        let mut buf = GenericArray::default();
+        left.serialize(&mut buf);
+        let deserialized_share = IndistinguishableHybridShare::<BA8, BA3>::deserialize(&buf);
+        assert_eq!(left, deserialized_share.unwrap());
+    }
+
+    #[test]
+    fn hybrid_add() {
+        let share1 = IndistinguishableHybridShare::<BA8, BA3> {
+            match_key: BA64::try_from(5).unwrap(),
+            value: BA3::try_from(1).unwrap(),
+            breakdown_key: BA8::try_from(2).unwrap(),
+        };
+
+        assert_eq!(
+            share1,
+            share1.clone() + IndistinguishableHybridShare::<BA8, BA3>::ZERO
+        );
+        // share 1: 5, 1, 2
+        // share 2: 2, 0, 3
+        // XOR
+        // expected: 7, 1, 1
+        let share2 = IndistinguishableHybridShare::<BA8, BA3> {
+            match_key: BA64::try_from(2).unwrap(),
+            value: BA3::try_from(0).unwrap(),
+            breakdown_key: BA8::try_from(3).unwrap(),
+        };
+
+        let expected = IndistinguishableHybridShare::<BA8, BA3> {
+            match_key: BA64::try_from(7).unwrap(),
+            value: BA3::try_from(1).unwrap(),
+            breakdown_key: BA8::try_from(1).unwrap(),
+        };
+
+        assert_eq!(share1 + share2, expected);
+    }
+
+    #[test]
+    fn test_shuffle_hybrid_inputs() {
+        const BATCHSIZE: usize = 50;
+        run(|| async {
+            let world = TestWorld::default();
+            let mut rng = world.rng();
+            let mut records = (0..BATCHSIZE)
+                .map(|_| TestIndistinguishableHybridReport {
+                    match_key: rng.gen::<u64>(),
+                    breakdown_key: rng.gen_range(0u32..1 << 8),
+                    value: rng.gen_range(0u32..1 << 3),
+                })
+                .collect::<Vec<_>>();
+
+            // let mut result: Vec<TestIndistinguishableHybridReport> = world
+            //     .semi_honest(records.clone().into_iter(), |ctx, input_rows| async move {
+            //         ctx.shuffle(input_rows).await.unwrap()
```

Comment:
This gives a compile error:
```
ctx.shuffle(input_rows).await.unwrap()
    |                         ^^^^^^^ expected `AdditiveShare<_>`, found `IndistinguishableHybridReport<_, _>`
 ```
 and I can't seem to find the right trait that accepts `I: IntoIterator<Item = Shuffleable> + Send` (instead of `I: IntoIterator<Item = AdditiveShare<S>> + Send`).

Reply:
That doesn't seem to be what's happening, as `ctx.shuffle` seems to be coming from [here](https://github.com/private-attribution/ipa/blob/main/ipa-core/src/protocol/ipa_prf/shuffle/mod.rs#L115). 

That said, it looks like this is going to change with #1402 (and even before that, this needs to bring in changes from main.) I'll sync with @andyleiserson tomorrow, but it looks like I may need to take a different approach with this to work with that PR.